src/main/kotlin/com/lykke/trade/volume/monitoring/service/entity/impl/TradeVolumeCacheImpl.kt

```diff
@@ -0,0 +1,136 @@
+
+    private fun getCumulativeVolumeForTradeVolume(volume: Volume, volumes: NavigableSet<Volume>): BigDecimal {
+        val higherVolume = volumes.higher(volume)
+        return if (higherVolume != null) {
+            cumulativeVolumeByTradeVolume[getVolumeKey(higherVolume)]!!.add(volume.volume)
+        } else {
+            volume.volume
+        }
+    }
+
+    private fun getTradeVolumesForPeriod(volume: Volume, volumes: NavigableSet<Volume>): List<Pair<Long, BigDecimal>> {
+        val result = ArrayList<Pair<Long, BigDecimal>>()
+        val volumesIterator = volumes.iterator()
+
+        updateCumulativeVolumes(volume, volumes)
+
+        while (volumesIterator.hasNext()) {
+            val currentVolume = volumesIterator.next()
+
+            val periodBoundVolume = volumes.floor(Volume(Integer.MAX_VALUE, Date(currentVolume.timestamp.time - cacheConfig.volumePeriod), BigDecimal.ZERO, volume.clientId, volume.assetId))
+            val volumeForPeriod = if (periodBoundVolume != null) {
+                cumulativeVolumeByTradeVolume[getVolumeKey(currentVolume)]!! - (cumulativeVolumeByTradeVolume[getVolumeKey(periodBoundVolume)]!! - periodBoundVolume.volume)
+            } else {
+                currentVolume.volume
+            }
+
+            result.add(currentVolume.timestamp.time to volumeForPeriod)
+
+            if (currentVolume == volume) {
+                break
+            }
+        }
+
+        return result
+    }
+
+    private fun updateCumulativeVolumes(volume: Volume, volumes: NavigableSet<Volume>) {
+        val volumesIterator = volumes.iterator()
+
+        while (volumesIterator.hasNext()) {
+            val currentVolume = volumesIterator.next()
+            if (currentVolume == volume) {
+                return
+            }
+            cumulativeVolumeByTradeVolume[getVolumeKey(currentVolume)] = cumulativeVolumeByTradeVolume[getVolumeKey(currentVolume)]!!.add(volume.volume)
+        }
+    }
+
+    private fun getLock(clientIdAssetId: String): Any {
+        lockByClientIdAssetId.putIfAbsent(clientIdAssetId, Object())
+        return lockByClientIdAssetId[clientIdAssetId] as Any
+    }
+
+    private fun getVolumeKey(volume: Volume): String {
+        return "${volume.tradeIdx}_${volume.timestamp.time}_${volume.clientId}_${volume.assetId}"
+    }
+
+    private fun getClientVolumesKey(clientId: String, assetId: String): String {
+        return "${clientId}_$assetId"
+    }
+
+    private fun getVolumes(clientIdAssetId: String): NavigableSet<Volume> {
+        return tradeVolumesByClientIdByAssetId
+                .getOrPut(clientIdAssetId) { TreeSet() }
+    }
+
+    private data class Volume(val tradeIdx: Int,
+                         val timestamp: Date,
+                         val volume: BigDecimal,
+                         val clientId: String,
+                         val assetId: String) : Comparable<Volume> {
+        override fun compareTo(other: Volume): Int {
+            return Comparator
```

Comment:
The comparator looks wrong:
1. second `reversed()` must be applied to `Comparator.comparingInt`
2. compare(other, this) -> compare(this, other)

What agreement about tradeIdx to pass it to the cache ? Just index of got trades from ME ?

Reply:
fixed, yes - trade id is `tradeIdx` is trade `index` from ME